internal/deploy/yaml.go

```diff
@@ -0,0 +1,213 @@
+}
+
+type Port struct {
+	Protocol   string `json:"protocol"`
+	Port       int    `json:"port"`
+	TargetPort int    `json:"targetPort"`
+}
+
+type Hooks struct {
+	Restart Restart `json:"restart"`
+}
+
+type Restart struct {
+	Before string `json:"before"`
+	After  string `json:"after"`
+}
+
+type CName struct {
+	DNSName string `json:"dnsName"`
+}
+
+var (
+	defaultVersion  = "v1"
+	defaultAppUnit  = 1
+	typeApplication = "Application"
+	typeJob         = "Job"
+
+	errEnvvarFormat = errors.New("environment variables should be in the format - name=value")
+)
+
+// GetChangeSetFromYaml reads an application.yaml file and returns a ChangeSet
+// from the file's values.
+func (o *Options) GetChangeSetFromYaml(filename string) (*ChangeSet, error) {
+	var application Application
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(b, &application)
+	if err != nil {
+		return nil, err
+	}
+
+	var envs []ketchv1.Env
+	if application.Environment != nil {
+		for _, env := range *application.Environment {
+			arr := strings.Split(env, "=")
+			if len(arr) != 2 {
+				return nil, errEnvvarFormat
+			}
+			envs = append(envs, ketchv1.Env{Name: arr[0], Value: arr[1]})
+		}
+	}
+	// processes, hooks, ports
+	var processes []ketchv1.ProcessSpec
+	var ketchYamlData ketchv1.KetchYamlData
+	if application.Processes != nil {
+		var beforeHooks []string
+		var afterHooks []string
+		ketchYamlProcessConfig := make(map[string]ketchv1.KetchYamlProcessConfig)
+		for _, process := range *application.Processes {
+			processes = append(processes, ketchv1.ProcessSpec{
+				Name:  process.Name,
+				Cmd:   strings.Split(process.Cmd, " "),
+				Units: process.Units,
+				Env:   envs,
+			})
+			if process.Hooks.Restart.Before != "" {
+				beforeHooks = append(beforeHooks, process.Hooks.Restart.Before)
+			}
+			if process.Hooks.Restart.After != "" {
+				afterHooks = append(afterHooks, process.Hooks.Restart.After)
+			}
+
+			var ports []ketchv1.KetchYamlProcessPortConfig
+			for _, port := range process.Ports {
+				ports = append(ports, ketchv1.KetchYamlProcessPortConfig{
+					Protocol:   port.Protocol,
+					Port:       port.Port,
+					TargetPort: port.TargetPort,
+				})
+			}
+			if len(process.Ports) > 0 {
+				ketchYamlProcessConfig[process.Name] = ketchv1.KetchYamlProcessConfig{
+					Ports: ports,
+				}
+			}
+		}
+
+		// assign hooks and ports (kubernetes processConfig) to ketch yaml data
+		// NOTE: there is a disparity in that the yaml file format implies that hooks and ports
+		// are per-process, while the AppSpec makes them per-deployment.
+		ketchYamlData = ketchv1.KetchYamlData{
+			Hooks: &ketchv1.KetchYamlHooks{
+				Restart: ketchv1.KetchYamlRestartHooks{
+					Before: beforeHooks,
+					After:  afterHooks,
+				},
+			},
+			Kubernetes: &ketchv1.KetchYamlKubernetesConfig{
+				Processes: ketchYamlProcessConfig,
+			},
+		}
+	}
+	c := &ChangeSet{
+		appName:              *application.Name,
```

Comment:
should we set `sourcePath`? 
do we support 
```bash
ketch app deploy app.yaml <source-dir>
```
?

I created the following app.yaml and I don't set `builder` because I know ketch provides a default one. 

```yaml
version: v1
type: Application
name: test
image: gcr.io/shipa-1000/sample-app:v1
framework: dev 
environment:
  - FOO=bar
appUnit: 2
cname:
  dnsName: test.10.10.10.20
```

Reply:
Good call. I added `sourcePath` to the `changeSet` and removed a check in `applyDefaults()` that added it, which I don't think made any sense.